### PR TITLE
SDL: improve linker settings for Apple platforms + v2 linter changes

### PR DESCRIPTION
**Pull request by kambala-decapitator**
Specify library name and version:  **sdl/2.0.20**

- shared library no longer enforces linked frameworks on consumers, as this info is already encoded in the library itself
- CoreHaptics appeared fairly recently (iOS 13 / macOS 10.15) and should be weakly linked to avoid "dylib not found" error when deploying to older OS versions
- also performed v2 migration based on linter report

Note that simple `-weak_framework CoreHaptics` doesn't work (as well as quoting it): it's translated into

    -weak_framework  -lCoreHaptics

Looks like something that should be adjusted in Conan.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by paulocoutinhox** [comment]
Hi. You need remove the unused imports:

image

See some warnings here:
https://github.com/conan-io/conan-center-index/pull/12472/files

File: recipes/sdl/all/conanfile.py
Comment (paulocoutinhox) [inline recipes/sdl/all/conanfile.py:4]: use other name for applying patchs because it is conflicting with the import name

**Review by paulocoutinhox — CHANGES_REQUESTED** [review]

**Comment by paulocoutinhox** [comment]
> Sorry, the build is only launched for Access Request users. You can request access [writing in this issue](https://github.com/conan-io/conan-center-index/issues/4).

@kambala-decapitator apply to you make PRs

File: recipes/sdl/all/conanfile.py
Comment (kambala-decapitator) [reply recipes/sdl/all/conanfile.py]: thanks! DIdn't notice that.

Hope I did it right, I'm not a python guy.

**Comment by kambala-decapitator** [comment]
> > Sorry, the build is only launched for Access Request users. You can request access [writing in this issue](https://github.com/conan-io/conan-center-index/issues/4).
> 
> @kambala-decapitator apply to you make PRs

yeah, I know, already requested access in that issue

**Comment by paulocoutinhox** [comment]
@madebr or @SpaceIm can you approve the user?

**Review by paulocoutinhox — DISMISSED** [review]
only need approve you to we check the build status

**Comment by madebr** [comment]
> Note that simple `-weak_framework CoreHaptics` doesn't work (as well as quoting it): it's translated into
> 
> ```
> -weak_framework  -lCoreHaptics
> ```
> 
> Looks like something that should be adjusted in Conan.

Perhaps, Conan needs something to add argument groups to flags.
e.g.:
```python
self.cpp_info.components["libsdl2"].exelinkflags.append(ArgGroup("-weak_framework", "CoreHaptics"))
```

e.g. in SDL's cmake config files, we ass `SHELL:` to avoid CMake treating every argument as a separate argument.
See [this sample] for example.
This avoids [CMake option de-duplication](https://cmake.org/cmake/help/latest/command/add_compile_options.html#option-de-duplication).

**Review by madebr — DISMISSED** [review]

**Comment by kambala-decapitator** [comment]
> Conan needs something to add argument groups to flags.

I think an alternative here could be adding `weak_frameworks` property (as `frameworks` one already exists) or extend `frameworks` to accept some special `AppleFramework` type that would have an `is_weak` property

**Comment by paulocoutinhox** [comment]
@madebr @SpaceIm add the @kambala-decapitator to the team pls

**Comment by madebr** [comment]
> @madebr @SpaceIm add the @kambala-decapitator to the team pls

That's not up to me.
@uilianries Can you please poke at the correct people that have approval rights?

File: recipes/sdl/all/conanfile.py
Comment (uilianries) [inline recipes/sdl/all/conanfile.py:3]: Need to update `required_conan_version = ">=1.45.0"` to `1.51.3`: https://docs.conan.io/en/latest/changelog.html#aug-2022

File: recipes/sdl/all/conanfile.py
Comment (uilianries) [inline recipes/sdl/all/conanfile.py:8]:
```suggestion
from conans import CMake
```
No longer used

File: recipes/sdl/all/conanfile.py
Comment (uilianries) [inline recipes/sdl/all/conanfile.py:5]:
```suggestion
from conan.tools.files import patch
```
Use the original name, it avoids confusion for who is used to read new recipes every day and know most of imports

File: recipes/sdl/all/conanfile.py
Comment (uilianries) [inline recipes/sdl/all/conanfile.py:184]:
```suggestion
        for data in self.conan_data.get("patches", {}).get(self.version, []):
```

File: recipes/sdl/all/conanfile.py
Comment (uilianries) [inline recipes/sdl/all/conanfile.py:185]:
```suggestion
            patch(**data)
```

File: recipes/sdl/all/conanfile.py
Comment (uilianries) [inline recipes/sdl/all/conanfile.py:369]:
```suggestion
        rm(self, "sdl2-config", os.path.join(self.package_folder, "bin"))
```
It's inverted: https://docs.conan.io/en/latest/reference/conanfile/tools/files/basic.html#conan-tools-files-rm

File: recipes/sdl/all/conanfile.py
Comment (kambala-decapitator) [reply recipes/sdl/all/conanfile.py:8]: it's still used: https://github.com/conan-io/conan-center-index/blob/6eaa24f5ca1c098cb6dd09d903d1407caa5f934b/recipes/sdl/all/conanfile.py#L358

**Comment by paulocoutinhox** [comment]
@kambala-decapitator do you want that i put this conanfile in my PR for sdl 2.24.0 to accelerate?

**Comment by kambala-decapitator** [comment]
> @kambala-decapitator do you want that i put this conanfile in my PR for sdl 2.24.0 to accelerate?

@paulocoutinhox personally I'm fine with that. But not sure if it's right from maintainers' POV.

**Comment by kambala-decapitator** [comment]
may I already trigger CI or PR approval is needed first?

**Comment by paulocoutinhox** [comment]
> may I already trigger CI or PR approval is needed first?

@uilianries can you help us with this?

**Comment by uilianries** [comment]
@kambala-decapitator please update the line 145: `self.requires("wayland/1.20.0")` -> `self.requires("wayland/1.21.0")`

There is a version conflict.

**Comment by kambala-decapitator** [comment]
> @kambala-decapitator please update the line 145: `self.requires("wayland/1.20.0")` -> `self.requires("wayland/1.21.0")`
> 
> There is a version conflict.

@uilianries done

File: recipes/sdl/all/conanfile.py
Comment (SpaceIm) [inline recipes/sdl/all/conanfile.py:5]:
```suggestion
from conan.tools.files import apply_conandata_patches, get, replace_in_file, rm, rmdir
```

File: recipes/sdl/all/conanfile.py
Comment (SpaceIm) [inline recipes/sdl/all/conanfile.py:185]:
```suggestion
        apply_conandata_patches(self)
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/sdl/all/conanfile.py
Comment (kambala-decapitator) [reply recipes/sdl/all/conanfile.py:184]: oh wow, wasn't aware of such function. Would be great to have to documented at https://github.com/conan-io/conan-center-index/blob/master/docs/v2_linter.md

**Review by paulocoutinhox — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**
